zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -2800,26 +2807,98 @@ public void setACL(final String path, List<ACL> acl, int version, StatCallback c
+            // If the next page still starts from czxidOffset, the children that not in the previous page
+            // but their offset is less than czxidOffset, they would be missed.
+            // HashSet is used to de-dup the duplicate children.
+            request.setCzxidOffset(isFirstPage ? czxidOffset : 0);
+            response = new GetChildrenPaginatedResponse();
+            ReplyHeader r = cnxn.submitRequest(h, request, response, wcb);
+            if (r.getErr() != 0) {
+                throw KeeperException.create(KeeperException.Code.get(r.getErr()),
+                        path);
+            }
+            minCzxId = response.getNextPageCzxid();
+            czxidOffset = response.getNextPageCzxidOffset();
+            needNextPage = needNextPage(maxReturned, minCzxId);
+
+            if (isFirstPage) {
+                // If all children are returned in the first page,
+                // no need to use hash set to de-dup children
+                if (!needNextPage) {
+                    updateNextPage(nextPage, minCzxId, czxidOffset);
+                    return response.getChildren();
+                }
+                children = new HashSet<>();
+                isFirstPage = false;
+            }
+
+            children.addAll(response.getChildren());
+        }
+
+        updateNextPage(nextPage, minCzxId, czxidOffset);
+
+        return new ArrayList<>(children);
+    }
+
+    /**
+     * Returns a list of all the children given the path.
+     * <p>
+     * The difference between this API and {@link #getChildren(String, boolean)} is:
+     * when there are lots of children and the network buffer exceeds {@code jute.maxbuffer},
+     * this API will fetch the children using pagination and be able to return all children;
+     * while {@link #getChildren(String, boolean)} will fail.
+     * <p>
+     * The final list of children returned is NOT strongly consistent with the server's data:
+     * the list might contain some deleted children if some children are deleted before the last page is fetched.
+     * <p>
+     * If the watch is true and the call is successful (no exception is thrown),
+     # a watch will be left on the node with the given path. The watch will be
+     * triggered by a successful operation that deletes the node of the given
+     * path or creates/deletes a child under the node.
+     *
+     * @param path the path of the parent node
+     * @param watch whether or not leave a watch on the given node
+     * @return a list of all children of the given path
+     * @throws KeeperException if the server signals an error with a non-zero error code.
+     * @throws InterruptedException if the server transaction is interrupted.
+     */
+    public List<String> getAllChildrenPaginated(String path, boolean watch)
+            throws KeeperException, InterruptedException {
+        return getChildren(path, watch ? watchManager.defaultWatcher : null, Integer.MAX_VALUE, 0L, 0, null);
```

Comment:
Discussed with JJ and Huizhi offline. The main idea is following:
1/ always going to offset 0 for same transaction Id approach is used to work around missing nodes from the same multi write (with same transaction id) when one of the node is deleted. The downside is that if the muti-write size is larger than 1 page, this would a dead loop.

2/ The new approach is that always use the last read node as the first for next batch. If the new batch does not have the last read node as the first node, we know there are deleted ones. We will revert back to offset 0 approach.